### PR TITLE
Add support for locating custom renderers

### PR DESCRIPTION
**Pull request by ncarrillo**
- What does the pull request do?

Adds support for custom renderers to override the defaults. For context, I'm working on a custom rendering backend for Avalonia (which includes a customized `IRenderer` impl) which I intend to have exist _out of tree_ to allow me to experiment freely.

- What is the current behavior?

`ImmediateRenderer` or `DeferredRenderer` are selected based on `useDeferred` passed to `UseWin32`

- What is the updated/expected behavior with this PR?

A custom renderer can potentially live outside of tree. 

- How was the solution implemented (if it's not obvious)?

Makes `Avalonia.Win32`'s `WindowImpl` try to locate another `IRenderer` before falling back to use either `ImmediateRenderer` or `DeferredRenderer`

The potentially controversial aspect is I made `TransformedBounds` public so that I can access it, I think the `ObsoleteAttribute` on it gives plenty of warning. 

Checklist:

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

### CODE REVIEW

**Comment by Gillibald** [comment]
Not sure if I like this approach of createing a custom renderer. Avalonia has a ServiceLocator and a valid renderer should be registered with it in my opinion instead of your Fun<x,y> approach. I much prefer a factory for this:

```
public interface IRendererFactory
{
IRenderer Create(IRenderRoot root, RendererOptions options);
}
```

**Comment by nc4rrillo** [comment]
I'm not opposed to that, I was currently just registering the Func, but I could add a factory. Went with the route of least resistance though :).

**Comment by nc4rrillo** [comment]
@Gillibald How does this look :-)

**Comment by Gillibald** [comment]
Looks good to me. I wonder if we can introduce a DefaultRendererFactory that returns either Immediate Renderer or Deferred Renderer depending on UseDeferredRendering

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 1]
```diff
@@ -124,7 +124,9 @@ public IScreenImpl Screen
         public IRenderer CreateRenderer(IRenderRoot root)
         {
             var loop = AvaloniaLocator.Current.GetService<IRenderLoop>();
-            return Win32Platform.UseDeferredRendering ?
+            var customRendererFactory = AvaloniaLocator.Current.GetService<IRendererFactory>();
+
+            return customRendererFactory != null ? customRendererFactory.Create(root, loop) : Win32Platform.UseDeferredRendering ?
```
Comment (grokys) [inline src/Windows/Avalonia.Win32/WindowImpl.cs]: This is a bit hard to read. Could you split it into separate `if` statements?

File: src/Avalonia.Visuals/Media/DrawingContext.cs [hunk 2]
```diff
@@ -59,7 +59,7 @@ private set
         //HACK: This is a temporary hack that is used in the render loop 
         //to update TransformedBounds property
         [Obsolete("HACK for render loop, don't use")]
-        internal Matrix CurrentContainerTransform => _currentContainerTransform;
+        public Matrix CurrentContainerTransform => _currentContainerTransform;
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/DrawingContext.cs]: Yeah... How to properly handle this is something we've been debating for like years... I'm fine with this hack for now personally, but we need to work out how we're going to do this properly.

**Review by grokys — CHANGES_REQUESTED** [review]
Looks good! Just a small formatting request.

**Comment by nc4rrillo** [comment]
@grokys This is ready for review again

**Review by grokys — APPROVED** [review]
Thanks @nc4rrillo !

**Pull request merged**